features/enlinkd/org.opennms.features.enlinkd.generator/src/main/java/org/opennms/enlinkd/generator/protocol/Protocol.java

```diff
@@ -0,0 +1,176 @@
+import org.opennms.netmgt.model.OnmsNode;
+import org.opennms.netmgt.model.OnmsSnmpInterface;
+import org.opennms.netmgt.model.monitoringLocations.OnmsMonitoringLocation;
+
+
+public abstract class Protocol<Element> {
+
+    final TopologySettings topologySettings;
+    final TopologyContext context;
+    private final RandomUtil random = new RandomUtil();
+
+    public Protocol(TopologySettings topologySettings, TopologyContext context) {
+        this.topologySettings = topologySettings;
+        this.context = context;
+    }
+
+    public void createAndPersistNetwork(){
+        this.context.currentProgress("%nCreating %s %s topology with %s Nodes, %s Elements, %s Links, %s SnmpInterfaces, %s IpInterfaces:",
+                topologySettings.getTopology(),
+                this.getProtocol(),
+                topologySettings.getAmountNodes(),
+                topologySettings.getAmountElements(),
+                topologySettings.getAmountElements(),
+                topologySettings.getAmountSnmpInterfaces(),
+                topologySettings.getAmountIpInterfaces());
+        OnmsCategory category = createCategory();
+        context.getTopologyPersister().persist(category);
+        List<OnmsNode> nodes = createNodes(topologySettings.getAmountNodes(), category);
+        context.getTopologyPersister().persist(nodes);
+
+        createAndPersistProtocolSpecificEntities(nodes);
+
+        List<OnmsSnmpInterface> snmpInterfaces = createSnmpInterfaces(nodes);
+        context.getTopologyPersister().persist(snmpInterfaces);
+        List<OnmsIpInterface> ipInterfaces = createIpInterfaces(snmpInterfaces);
+        context.getTopologyPersister().persist(ipInterfaces);
+    }
+
+    protected abstract void createAndPersistProtocolSpecificEntities(List<OnmsNode> nodes);
+
+    protected abstract TopologyGenerator.Protocol getProtocol();
+
+    private OnmsCategory createCategory() {
+        OnmsCategory category = new OnmsCategory();
+        category.setName(TopologyGenerator.CATEGORY_NAME);
+        return category;
+    }
+
+    private OnmsMonitoringLocation createMonitoringLocation() {
+        OnmsMonitoringLocation location = new OnmsMonitoringLocation();
+        location.setLocationName("Default");
+        location.setMonitoringArea("localhost");
+        return location;
+    }
+
+    protected List<OnmsNode> createNodes(int amountNodes, OnmsCategory category) {
+        OnmsMonitoringLocation location = createMonitoringLocation();
+        ArrayList<OnmsNode> nodes = new ArrayList<>();
+        for (int i = 0; i < amountNodes; i++) {
+            nodes.add(createNode(i, location, category));
+        }
+        return nodes;
+    }
+
+    protected OnmsNode createNode(int count, OnmsMonitoringLocation location, OnmsCategory category) {
+        OnmsNode node = new OnmsNode();
+        node.setId(count); // we assume we have an empty database and can just generate the ids
+        node.setLabel("Node" + count);
```

Comment:
The resulting nodes are not visible from the UI when viewing the node list - running the following SQL statemen makes them visible: `update nodes set nodetype = 'A';`.

Can we make sure they are visible by default?

Reply:
done.